### PR TITLE
copr: basic DAG expression evaluation support

### PR DESCRIPTION
**Pull request by andelf**
TODO:

- [ ] fn sig verification
- [x] doc comment

### CODE REVIEW

File: src/coprocessor/dag/expr.rs [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+            _ => Some(self.f64()),
+        }
+    }
+}
+
+impl Expression {
+    fn eval_int(&self, row: &[Datum], ctx: &StatementContext) -> Result<Option<i64>> {
+        match *self {
+            Expression::Constant(ref datum) => Ok(datum.to_i64()),
+            Expression::ColumnRef(offset) => row.get(offset).map(Datum::to_i64).ok_or_else(|| Error::Other("offset not fount")),
```
Comment (BusyJay) [inline src/coprocessor/dag/expr.rs]: Should state which offset not found.

File: src/coprocessor/dag/expr.rs [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+                    ScalarFuncSig::CastRealAsInt =>
+                        fun.children[0].eval_real(row, ctx).map(|v| v.map(|v| v as i64)),
+                    //
+                    ScalarFuncSig::LTInt => {
+                        let lhs = try!(fun.children[0].eval_int(row, ctx));
+                        let rhs = try!(fun.children[1].eval_int(row, ctx));
+                        match (lhs, rhs) {
+                            (None, _) => Ok(None),
+                            (_, None) => Ok(None),
+                            (Some(l), Some(r)) => Ok(Some((l < r) as i64))
```
Comment (BusyJay) [inline src/coprocessor/dag/expr.rs]: Match this first, then you can use `_` for others.

File: src/coprocessor/dag/expr.rs [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+            from()
+            description("io error")
+            display("I/O error: {}", err)
+            cause(err)
+        }
+        Type { has: &'static str, expected: &'static str } {
+            description("type error")
+            display("type error: cannot get {:?} result from {:?} expression", expected, has)
+        }
+        Codec(err: util::codec::Error) {
```
Comment (siddontang) [inline src/coprocessor/dag/expr.rs]: remove `util::`

Comment (andelf) [inline src/coprocessor/dag/expr.rs]: there's 2 codec mod.

File: src/coprocessor/dag/expr.rs [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+
+use chrono::FixedOffset;
+
+use tipb::expression::{ExprType, ScalarFuncSig};
+use tipb::expression::Expr;
+use tipb::select::DAGRequest;
+
+use util::codec::number::{NumberDecoder, NumberEncoder};
+use coprocessor::codec::mysql::decimal::DecimalDecoder;
+use coprocessor::codec::mysql::MAX_FSP;
```
Comment (hicqu) [inline src/coprocessor/dag/expr.rs]: Module `copressor` and `super::super` are the same module. Should we put them together?

File: src/coprocessor/dag/expr.rs [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+
+impl Expression {
+    fn eval_int(&self, row: &[Datum], ctx: &StatementContext) -> Result<Option<i64>> {
+        match *self {
+            Expression::Constant(ref datum) => Ok(datum.to_i64()),
+            Expression::ColumnRef(offset) => {
+                row.get(offset).map(Datum::to_i64).ok_or_else(|| Error::ColumnOffset(offset))
+            }
+            Expression::ScalarFn(ref fun) => {
+                match fun.sig {
```
Comment (hicqu) [inline src/coprocessor/dag/expr.rs]: Shoud we move this to a public place? `eval_real` and other `eval_*` will use this, too.

**Comment by hicqu** [comment]
LGTM

File: src/coprocessor/dag/expr/mod.rs [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+pub struct FnCall {
+    sig: ScalarFuncSig,
+    children: Vec<Expression>,
+}
+
+impl Datum {
+    fn to_i64(&self) -> Option<i64> {
+        match *self {
+            Datum::Null => None,
+            _ => Some(self.i64()),
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/mod.rs]: This may not be the correct way to convert `Datum` to `i64`.

**Comment by AndreMouche** [comment]
There're too many compile warnings in your branch, try to fix it.

File: src/coprocessor/dag/expr/mod.rs [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+
+/// A single scalar function call
+#[derive(Debug, Clone, PartialEq)]
+pub struct FnCall {
+    sig: ScalarFuncSig,
+    children: Vec<Expression>,
+}
+
+impl Datum {
+    fn to_i64(&self, field_type: &FieldType) -> Option<i64> {
```
Comment (hicqu) [inline src/coprocessor/dag/expr/mod.rs]: should return `Result<Option<i64>>`?

**Pull request merged**